sdk/trace/batch_span_processor_test.go

```diff
@@ -633,3 +643,236 @@ func TestBatchSpanProcessorConcurrentSafe(t *testing.T) {
+		wantProcessedDataPoints = append(wantProcessedDataPoints, metricdata.DataPoint[int64]{
+			Value: expectation.queueFullProcessed,
+			Attributes: attribute.NewSet(
+				semconv.OTelComponentTypeBatchingSpanProcessor,
+				componentNameAttr,
+				semconv.ErrorTypeKey.String(string(queueFull)),
+			),
+		})
+	}
+
+	// sm.Metrics also includes otel.sdk.span.live and otel.sdk.span.started
+	if len(wantProcessedDataPoints) > 0 {
+		require.Len(t, sm.Metrics, 5)
+	} else {
+		require.Len(t, sm.Metrics, 4)
+	}
+
+	baseAttrs := attribute.NewSet(
+		semconv.OTelComponentTypeBatchingSpanProcessor,
+		componentNameAttr,
+	)
+
+	want := metricdata.Metrics{
+		Name:        otelconv.SDKProcessorSpanQueueCapacity{}.Name(),
+		Description: otelconv.SDKProcessorSpanQueueCapacity{}.Description(),
+		Unit:        otelconv.SDKProcessorSpanQueueCapacity{}.Unit(),
+		Data: metricdata.Sum[int64]{
+			DataPoints:  []metricdata.DataPoint[int64]{{Attributes: baseAttrs, Value: expectation.queueCapacity}},
+			Temporality: metricdata.CumulativeTemporality,
+			IsMonotonic: false,
+		},
+	}
+	metricdatatest.AssertEqual(t, want, sm.Metrics[0], metricdatatest.IgnoreTimestamp())
+
+	want = metricdata.Metrics{
+		Name:        otelconv.SDKProcessorSpanQueueSize{}.Name(),
+		Description: otelconv.SDKProcessorSpanQueueSize{}.Description(),
+		Unit:        otelconv.SDKProcessorSpanQueueSize{}.Unit(),
+		Data: metricdata.Sum[int64]{
+			DataPoints:  []metricdata.DataPoint[int64]{{Attributes: baseAttrs, Value: expectation.queueSize}},
+			Temporality: metricdata.CumulativeTemporality,
+			IsMonotonic: false,
+		},
+	}
+	metricdatatest.AssertEqual(t, want, sm.Metrics[1], metricdatatest.IgnoreTimestamp())
+
+	if len(wantProcessedDataPoints) > 0 {
+		want = metricdata.Metrics{
+			Name:        otelconv.SDKProcessorSpanProcessed{}.Name(),
+			Description: otelconv.SDKProcessorSpanProcessed{}.Description(),
+			Unit:        otelconv.SDKProcessorSpanProcessed{}.Unit(),
+			Data: metricdata.Sum[int64]{
+				DataPoints:  wantProcessedDataPoints,
+				Temporality: metricdata.CumulativeTemporality,
+				IsMonotonic: true,
+			},
+		}
+		metricdatatest.AssertEqual(t, want, sm.Metrics[2], metricdatatest.IgnoreTimestamp())
+	}
+}
+
+// blockingExporter blocks until the exported span is removed from the channel.
+type blockingExporter struct {
+	shutdown chan (struct{})
+	total    atomic.Int32
+}
+
+func newBlockingExporter() *blockingExporter {
+	e := &blockingExporter{shutdown: make(chan struct{})}
+	return e
+}
+
+func (e *blockingExporter) Shutdown(ctx context.Context) error {
+	select {
+	case <-e.shutdown:
+	default:
+		close(e.shutdown)
+	}
+	return ctx.Err()
+}
+
+func (e *blockingExporter) ExportSpans(ctx context.Context, s []ReadOnlySpan) error {
+	e.total.Add(int32(len(s)))
+	<-e.shutdown
+	return ctx.Err()
+}
+
+func (e *blockingExporter) waitForSpans(n int32) {
+	// Wait for all n spans to reach the export call
+	// nolint: revive // intentionally empty block
```

Comment:
How about calling https://pkg.go.dev/runtime#Gosched inside the loop so that it does not starve other goroutines?

Reply:
done